### PR TITLE
Add abstraction practice

### PR DESCRIPTION
**Pull request by rosariogarcia**
Hi Carlos, I uploaded my practice of abstraction and interface concepts.

Order: calculate area and perimeter for square, rectangle and circle

### CODE REVIEW

File: src/main/java/com/fundacionjala/geometry/Circle.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    public Circle(double radio) {
+        this.radio = radio;
+    }
+
+    public double calculateArea() {
+        return Math.PI * Math.pow(radio, 2);
+    }
+
+    public double calculatePerimeter() {
+        return 2 * Math.PI * radio;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Circle.java]: Magic number

File: src/main/java/com/fundacionjala/geometry/Constants.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.fundacionjala.geometry;
+
+/**
+ * Created by RosarioGarcia on 6/16/2016.
+ */
+public class Constants {
+
+    public static final int DELTA = 0;
+
+    private Constants(){
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Constants.java]: space before braces

``` java
private Constants() {
```

File: src/main/java/com/fundacionjala/geometry/Rectangle.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        this.base = base;
+        this.high = high;
+    }
+
+    public double calculateArea() {
+        return base * high;
+    }
+
+    public double calculatePerimeter() {
+        return 2 * base + 2 * high;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Rectangle.java]: magic number

File: src/test/java/com/fundacionjala/geometry/CircleTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by RosarioGarcia on 6/16/2016.
+ */
+public class CircleTest {
+
+    private static double radio = 2;
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/CircleTest.java]: This should be a constant

``` java
private static final double RADIO = 2;
```

File: src/test/java/com/fundacionjala/geometry/CircleTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private Shape circle;
+
+    @Before
+    public void serUp() {
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/CircleTest.java]: Typo
setUp

File: src/main/java/com/fundacionjala/geometry/Rectangle.java (hunk 3 above)
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Rectangle.java]: Optimize the formula

``` java
return FACTOR_FORMULA * (base + high);
```

**Comment by rosariogarcia** [comment]
Hi @carledriss, I updated my task with the last observations in classes.

File: src/main/java/com/fundacionjala/geometry/Circle.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.fundacionjala.geometry;
+
+import static java.lang.Math.PI;
+import static java.lang.Math.pow;
+
+/**
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Circle.java:6]: This should be

``` java
/**
 * Class to calculate area and perimeter of Circle
 *
 * @author Rosario Garcia
 */
```

Please update all your classes

File: src/test/java/com/fundacionjala/geometry/CircleTest.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    private Shape circle;
+
+    @Before
+    public void setUp() {
+        circle = new Circle(RADIO);
+    }
+
+    @Test
+    public void testCalculateCircleArea() {
+        double expectedArea = PI * pow(RADIO, Constants.FACTOR_FORMULA);
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/CircleTest.java]: This variable should final

File: src/test/java/com/fundacionjala/geometry/CircleTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        assertEquals("Area is not equal", expectedArea, circle.calculateArea(), Constants.DELTA);
+    }
+
+    @Test
+    public void testCalculateCirclePerimeter() {
+        double expectedPerimeter = Constants.FACTOR_FORMULA * RADIO * PI;
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/CircleTest.java]: This variable should final

File: src/test/java/com/fundacionjala/geometry/CircleTest.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        assertEquals("Perimeter is not equal", expectedPerimeter, circle.calculatePerimeter(), Constants.DELTA);
+    }
+
+    @Test
+    public void testCalculateAreaWhenRadioIsZero() {
+        circle = new Circle(RADIO_ZERO);
+        int expectedValue = 0;
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/CircleTest.java]: This variable should final

File: src/test/java/com/fundacionjala/geometry/CircleTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        assertEquals("Area is not zero", expectedValue, circle.calculateArea(), Constants.DELTA);
+    }
+
+    @Test
+    public void testCalculatePerimeterWhenRadioIsZeroShouldBePI() {
+        circle = new Circle(RADIO_ZERO);
+        int expectedValue = 0;
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/CircleTest.java]: This variable should final

File: src/test/java/com/fundacionjala/geometry/RectangleTest.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    private Shape rectangle;
+
+    @Before
+    public void setUp() {
+        rectangle = new Rectangle(BASE, HIGH);
+    }
+
+    @Test
+    public void testCalculateRectanglePerimeter() {
+        int expectedPerimeter = 12;
```
Comment (carledriss) [inline src/test/java/com/fundacionjala/geometry/RectangleTest.java]: This variable should final

**Comment by carledriss** [comment]
@rosariogarcia 
After minor changes
Looks good to me

File: src/main/java/com/fundacionjala/geometry/Circle.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+    public Circle(double radio) {
+        this.radio = radio;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public double calculateArea() {
+        return PI * pow(radio, Constants.FACTOR_FORMULA);
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Circle.java:23]: Please add a static import

**Comment by rosariogarcia** [comment]
Hi @carledriss, I finished to correct your last observations

File: src/main/java/com/fundacionjala/geometry/Constants.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package com.fundacionjala.geometry;
+
+/**
+ * Class to define constants used in several classes
+ *
+ * @author Rosario Garcia
+ */
+public class Constants {
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/geometry/Constants.java:8]: Don't forget to add a private constructor for constant class (static methods and attributes)

**Pull request merged**